plugins/ril.c

```diff
@@ -77,83 +81,162 @@ struct ril_data {
+			__func__, ril_radio_state_to_string(radio_state),
+			rd->init_state, rd->ofono_online);
+
+	rd->radio_state = radio_state;
+
+	/*
+	 * Before showing the modem as powered we make sure the radio is off so
+	 * we start in a sane state (as in AOSP). Note that we must always
+	 * receive an event with the current radio state on initialization
+	 * and also that power state has changed effectively when the event with
+	 * the new radio state has been received (we cannot rely on the reply to
+	 * RIL_REQUEST_RADIO_POWER). Finally, note that powering off on start is
+	 * a must for turbo devices (otherwise radio state never moves from
+	 * unavailable).
+	 */
+	if (rd->init_state) {
+		if (radio_state != RADIO_STATE_OFF) {
+			DBG("powering off radio on init");
 
-		ofono_info("%s: state: %s rd->ofono_online: %d",
-				__func__,
-				ril_radio_state_to_string(radio_state),
-				rd->ofono_online);
+			ril_send_power(rd, FALSE, ril_set_powered_off_cb, modem);
+		} else {
+			DBG("calling set_powered(TRUE)");
+			rd->init_state = FALSE;
+
+			/* Note that hw is powered, but offline */
+			ofono_modem_set_powered(modem, TRUE);
+		}
+		return;
+	}
+
+	switch (radio_state) {
+	case RADIO_STATE_ON:
 
-		rd->radio_state = radio_state;
-
-		switch (radio_state) {
-		case RADIO_STATE_ON:
-
-			if (rd->radio_settings == NULL) {
-				char *rs_driver;
-				struct ril_radio_settings_driver_data
-						rs_data = { rd->ril, modem };
-
-				if (rd->vendor == OFONO_RIL_VENDOR_QCOM_MSIM)
-					rs_driver = QCOMMSIMMODEM;
-				else
-					rs_driver = RILMODEM;
-
-				rd->radio_settings =
-					ofono_radio_settings_create(modem,
-							rd->vendor, rs_driver,
-							&rs_data);
-			}
-
-			break;
-
-		case RADIO_STATE_UNAVAILABLE:
-		case RADIO_STATE_OFF:
-
-			/*
-			 * Unexpected radio state change, as we are supposed to
-			 * be online. UNAVAILABLE has been seen occassionally
-			 * when powering off the phone. We wait 5 secs to avoid
-			 * too fast re-spawns, then exit with error to make
-			 * upstart re-start ofono.
-			 */
-			if (rd->ofono_online) {
-				ofono_error("%s: radio self-powered off!",
-						__func__);
-				sleep(5);
-				exit(1);
-			}
-			break;
-		default:
-			/* Malformed parcel; no radio state == broken rild */
-			g_assert(FALSE);
+		if (rd->radio_settings == NULL) {
+			struct ril_radio_settings_driver_data
+					rs_data = { rd->ril, modem };
+
+			rd->radio_settings =
+				ofono_radio_settings_create(modem,
+					rd->vendor,
+					get_driver_type(rd,
+						OFONO_ATOM_TYPE_RADIO_SETTINGS),
+					&rs_data);
 		}
+
+		break;
+	case RADIO_STATE_UNAVAILABLE:
+	case RADIO_STATE_OFF:
+		/*
+		 * Unexpected radio state change, as we are supposed to
+		 * be online. UNAVAILABLE has been seen occassionally
+		 * when powering off the phone. In midori there are glitches
+		 * when entering SIM PIN (ON->OFF->ON). We simply ignore
+		 * the event.
+		 */
+		if (rd->ofono_online)
+			ofono_info("%s: radio self-powered off!",
```

Comment:
This is a major change in semantics...  So what happens if the radio powers off and just stays like that?